bigtable/tests/data_integration_test.cc

```diff
@@ -30,66 +92,231 @@ int main(int argc, char* argv[]) try {
-                                  {bigtable::SetCell(family, "c0", 0, "v2"),
-                                   bigtable::SetCell(family, "c1", 0, "v3")}),
-  };
-  table.BulkApply(std::move(bulk));
-  std::cout << "bulk mutation successful" << std::endl;
+  return RUN_ALL_TESTS();
 
-  return 0;
 } catch (bigtable::PermanentMutationFailure const& ex) {
   std::cerr << "bigtable::PermanentMutationFailure raised: " << ex.what()
             << " - " << ex.status().error_message() << " ["
             << ex.status().error_code()
             << "], details=" << ex.status().error_details() << std::endl;
+  int count = 0;
+  for (auto const& failure : ex.failures()) {
+    std::cerr << "failure[" << count++
+              << "] {key=" << failure.mutation().row_key() << "}" << std::endl;
+  }
   return 1;
 } catch (std::exception const& ex) {
   std::cerr << "Standard exception raised: " << ex.what() << std::endl;
   return 1;
+} catch (...) {
+  std::cerr << "Unknown exception raised." << std::endl;
+  return 1;
+}
+
+namespace {
+/**
+ * Compare two cells, think about the spaceship operator.
+ *
+ * @return `< 0` if lhs < rhs, `0` if lhs == rhs, and `> 0` otherwise.
+ */
+int CellCompare(bigtable::Cell const& lhs, bigtable::Cell const& rhs);
+
+/**
+ * Compare two sets of cells.
+ *
+ * Unordered because ReadRows does not guarantee a particular order.
+ */
+void CheckEqualUnordered(std::vector<bigtable::Cell> expected,
+                         std::vector<bigtable::Cell> actual);
+
+}  // anonymous namespace
+
+namespace bigtable {
+//@{
+/// @name Helpers for GTest.
+bool operator==(Cell const& lhs, Cell const& rhs) {
+  return CellCompare(lhs, rhs) == 0;
+}
+
+bool operator<(Cell const& lhs, Cell const& rhs) {
+  return CellCompare(lhs, rhs) < 0;
+}
+
+/**
+ * This function is not used in this file, but it is used by GoogleTest; without
+ * it, failing tests will output binary blobs instead of human-readable text.
+ */
+void PrintTo(bigtable::Cell const& cell, std::ostream* os) {
+  *os << "  row_key=" << cell.row_key() << ", family=" << cell.family_name()
+      << ", column=" << cell.column_qualifier()
+      << ", timestamp=" << cell.timestamp() << ", value=" << cell.value()
+      << ", labels={" << absl::StrJoin(cell.labels(), ",") << "}";
+}
+//@}
+}  // namespace bigtable
+
+TEST_F(DataIntegrationTest, TableApply) {
+  auto table = CreateTable();
+  std::string const row_key = "row-key-1";
+  std::vector<bigtable::Cell> created{
+      {row_key, family, "c0", 1000, "v1000", {}},
+      {row_key, family, "c1", 2000, "v2000", {}}};
+  Apply(*table, row_key, created);
+
+  std::vector<bigtable::Cell> expected{
+      {row_key, family, "c0", 1000, "v1000", {}},
+      {row_key, family, "c1", 2000, "v2000", {}}};
+
+  auto actual = ReadRows(*table, bigtable::Filter::PassAllFilter());
+  CheckEqualUnordered(expected, actual);
+
+  DeleteTable();
+}
+
+TEST_F(DataIntegrationTest, TableBulkApply) {
+  auto table = CreateTable();
```

Comment:
Please move `CreateTable()` and `DeleteTable()` to the `Setup()` and `TearDown()` functions in the test fixture.

Reply:
Also, if we can use different table names for each test then the tests could run in parallel.  It may not be possible to do that with `TearDown()` and `SetUp()`, if it is not, then having different table names is more interesting / important..

Reply:
Yes, TearDown() and SetUp() will not be possible as we keep different table names for different tests.
would creating a class different then table fixture (e.g. class TableIntegrationTest : public ::testing::Test ) be a good idea?
Currently table fixture is making few assumption like table name, instance name (saw some hard coding) which might not be required for integration test, a new class can help all 3 integration tests.

Reply:
> Yes, TearDown() and SetUp() will not be possible as we keep different table names for
> different tests.

That is fine.  Then use different table names in `TEST_F(DataIntegrationTest, TableBulkApply)` and `TEST_F(DataIntegrationTest, TableApply)`.

> would creating a class different then table fixture (e.g. 
> `class TableIntegrationTest : public ::testing::Test` ) be a good idea?
> Currently table fixture is making few assumption like table name, instance name

Yes.

> (saw some hard coding)

Can you let me know where?  If you mean the table ids picked in `filters_integration_test.cc` those are fine.  Look at the tables as disposable in this context, as long as each test uses a different name things are Okay.

> which might not be required for integration test, a new class can help all 3 integration tests.

I have some trouble parsing your proposal.  Can you explain further?  Maybe a small doc or email, or I am happy to join a GVC if that helps (assuming timezone issues can be worked out).

What I am proposing is that you refactor the `FilterIntegrationTest` and `FilterTestEnvironment` classes to the `bigtable/client/testing` directory (you will need to change their names, of course).  Use those classes in `filters_integration_test.cc`, and `data_integration_test.cc`.  That would require you to change `TEST_F(DataIntegrationTest, TableBulkApply)` and `TEST_F(DataIntegrationTest, TableApply)` to use a different table id in each test.

Does that make sense?

Reply:
added TableIntegrationTest and TableTestEnvironment class and all the integration test can use it. Copied common functions of data_integration_test and filter_integration_test to TableIntegrationTest